feat(max): conversation history

skoob13 opened the pull request:
## Problem

Customers want to see their previous conversations with Max. This PR brings the UI for chat history.

## Changes

- Add UI for history: the most recent conversations and a list of all conversations.
- Allow users to open and continue the conversation from where they left off.
- Handle conversations that are in progress when someone refreshes the page.
- Preserve the opened conversation in the URL and restore it on a fresh start.

### Demo

<img width="561" alt="Screenshot 2025-04-29 at 19 55 37" src="https://github.com/user-attachments/assets/d1b441db-0494-4b8b-b999-28a4311e60d2" /> <img width="559" alt="Screenshot 2025-04-29 at 20 42 13" src="https://github.com/user-attachments/assets/514e29a8-1099-4268-a17c-c7a2d7e9bf2f" />

### Next

Found two edge cases:
- Chat history needs virtualization because the conversation list endpoint can be slow on scale.
- If you send a message, refresh the page, and open the conversation, the initial message doesn't show. This is because of the internal implementation of the LangGraph's checkpointer, which marks the initial state as retrievable after the first run has finished. This might be fixed by writing the state before executing.

## Does this work well for both Cloud and self-hosted?

No

## How did you test this code?

Unit tests, stories, and manual tests.

skoob13 committed:
feat(conversation): add conversation types and status

- Introduced `ConversationType` and `ConversationStatus` enums to categorize conversation states.
- Updated `Conversation` interface to include new properties: `type`, `status`, and `title`.
- Enhanced `ConversationDetail` interface to include an array of `AssistantMessage`.
- Added API methods to list and get conversation details, returning `ConversationDetail` type.

skoob13 committed:
feat: logic

skoob13 committed:
feat: history

skoob13 committed:
feat: use conversation id

skoob13 committed:
feat(conversation): refactor conversation handling and improve serialization

- Replaced `ConversationSerializer` with `ConversationDetailSerializer` in `ConversationViewSet`.
- Updated conversation serialization in the `Assistant` class to use `ConversationSerializer`.
- Enhanced conversation filtering in `ConversationViewSet` to order by `updated_at`.
- Removed unused `ConversationType` enum and related properties from the `Conversation` interface.
- Introduced utility function `getConversationUrl` for generating conversation URLs in the frontend.
- Updated conversation history components to utilize the new URL structure.

skoob13 committed:
feat: skeleton

skoob13 committed:
feat(conversation): refactor conversation date handling and remove unused logic

- Replaced `humanFriendlyDuration` with `formatConversationDate` for better date formatting in `ConversationHistory` and `HistoryPreview` components.
- Removed `maxHistoryLogic.ts` as it was no longer needed.
- Updated `scrollThreadToBottom` to accept a behavior parameter for smoother scrolling options.
- Enhanced `loadConversationHistory` to conditionally update the current thread based on a new parameter.

skoob13 committed:
feat: shared filters

skoob13 committed:
fix: date formatting

skoob13 committed:
fix: sync scrolling

skoob13 committed:
feat: update QuestionInput to support floating state

- Added `isFloating` prop to `QuestionInput` component for conditional rendering.
- Updated `MaxInstance` to pass `isFloating` prop based on thread state.

skoob13 committed:
feat: data states

skoob13 committed:
fix: loaders and headers

skoob13 committed:
feat(conversation): add tests for conversation ordering and message filtering

- Implemented a test to verify that conversations are listed in order of their last updated timestamp.
- Added a test to ensure the Assistant class correctly filters messages based on content and UI payload.
- Updated frontend storybook to handle various conversation states, including loading and empty states.

skoob13 committed:
fix: types

skoob13 committed:
fix: tests

skoob13 committed:
fix: tests

skoob13 committed:
Merge branch 'master' into feat/max-conversations-ui

skoob13 committed:
fix: weird ts error

Twixes committed:
Fix alignment of "Open as main focus" tooltip

Twixes committed:
Reduce font size of the side panel title app-wide

Twixes commented on ee/hogai/utils/filters.py:
Because `filters` is so overloaded as a term in PostHog: did you consider moving this func into `ee/hogai/utils/helpers.py`? (I know "helpers" isn't specific either, but that already contains `filter_and_merge_messages()`)

Twixes commented on frontend/src/scenes/max/Max.tsx:
```diff
@@ -95,15 +100,21 @@ export function MaxInstance({ sidePanel }: MaxInstanceProps): JSX.Element {
             {sidePanel && (
-                <SidePanelPaneHeader>
+                <SidePanelPaneHeader title={chatTitle}>
                     <LemonButton
                         size="small"
                         sideIcon={<IconPlus />}
                         onClick={() => startNewConversation()}
                         tooltip="Start a new chat"
                         tooltipPlacement="bottom"
                     />
-                    <div className="flex-1" />
+                    <LemonButton
+                        size="small"
+                        sideIcon={<IconClockRewind />}
+                        onClick={() => toggleConversationHistory()}
+                        tooltip="Open chat history"
+                        tooltipPlacement="bottom"
+                    />
```
Something I'm missing here: a "back" button. While the home view technically is a new chat, it doesn't feel that way – I only feel like I'm in a new chat once I actually _start_ a chat.

Here's my user story: I just opened up Max and went into one of my past chats. From that point, I expected to be able to go back to the home view – but I can only do that by starting a new chat, which didn't feel quite intuitive.

![2025-04-30 18 37 41](https://github.com/user-attachments/assets/e6d33b70-72fa-4acd-8414-d478e6c4c9c4)

skoob13 replied:
Fair feedback. I felt the same way in the modern assistant interfaces but got used to it, so I expect to have the "new conversation" button somewhere. I made two additional concepts (UI is not final).

#### Back button and the history button

The navigation between screens is intuitive, but quite distant from the pattern. Also, the behavior of the history button is not clear. For example, when we open a chat from a URL, should we go back to the history or home screen?

![2025-05-01 17 42 01](https://github.com/user-attachments/assets/32627377-154e-4832-beb5-8c7997e62c84)

#### Remove the history button and add the back button

Max opens in a new chat mode by default. When you press the back button, you go to the history of conversations. If you're in a chat and press the back button, you go back to the history. I like this concept more than the previous one because the navigation is straightforward, but I like less that the home screen is the chat history.

![2025-05-01 18 09 26](https://github.com/user-attachments/assets/cb5aba08-5537-4f04-9ff8-c283a881771c)

What do you think?

Twixes reviewed (changes requested):
Leaving a few notes from my experience and partial code review so far, more to come as   there's plenty of code here

Twixes commented:
Also pushed two small changes:
1. [Minor tooltip placement fixes](https://github.com/PostHog/posthog/pull/31706/commits/3617479c8a445df5d3dc2fef0860247df9fdb4d5)
2. [Change the side panel title size from `text-base` to `text-sm` for all side panels](https://github.com/PostHog/posthog/pull/31706/commits/b08173444eb9f17f751eb86d4f0a5a3facd6ffba) – this actually looks better in general, but critically for max, `text-base` is too large for showing any meaningful chat title without overflowing

skoob13 commented on frontend/src/scenes/max/maxLogic.tsx:
```diff
@@ -474,7 +626,49 @@ export const maxLogic = kea<maxLogicType>([
+
+                // Existing conversation
+                if (conversation) {
+                    return conversation.title ?? 'Chat'
+                }
+
+                // Conversation is loading
+                if (conversationId) {
+                    return ''
+                }
+
+                return 'New chat'
```
Fixed. When a new conversation receives a title, it will now be updated.

skoob13 committed:
fix: titles

skoob13 commented on ee/hogai/api/serializers.py:
```diff
@@ -0,0 +1,34 @@
+_conversation_fields = ["id", "status", "title", "created_at", "updated_at"]
+
+
+class ConversationSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Conversation
+        fields = _conversation_fields
+        read_only_fields = fields
+
+
+class ConversationDetailSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Conversation
+        fields = [*_conversation_fields, "messages"]
+        read_only_fields = fields
+
+    messages = serializers.SerializerMethodField()
+
+    def get_messages(self, conversation: Conversation):
+        graph: CompiledStateGraph = self.context["assistant_graph"]
+        snapshot = graph.get_state({"configurable": {"thread_id": str(conversation.id)}})
+        try:
+            state = AssistantState.model_validate(snapshot.values)
+            return [message.model_dump() for message in state.messages if should_output_assistant_message(message)]
+        except (pydantic.ValidationError, KeyError):
+            return []
```
Thank you! I didn't know it.

skoob13 committed:
fix: rename serializers

skoob13 committed:
fix: remove filters

skoob13 committed:
fix: storybook

skoob13 committed:
fix: scrolling after opening/closing the history

skoob13 committed:
feat: back btn

skoob13 committed:
feat: wip second back button option

Twixes commented on frontend/src/scenes/max/HistoryPreview.tsx:
```diff
@@ -0,0 +1,43 @@
+
+    return (
+        <div className="max-w-120 w-full self-center flex flex-col gap-2">
+            <h3 className="text-sm font-medium text-secondary">Past chats</h3>
```
So these are everyone's chats from the environment, not just yours, right? Chat "authorship" isn't surfaced anywhere, so I think people will be surprised to see chats of other people without any signal that these indeed belong to other people in the same org.
Maybe we can show the chat "author" next to each position in this list? Or, only show chats of the current user.

skoob13 replied:
Chat history only displays your chats now. I don't think we should show chats from the environment, and it'd be confusing.

Twixes commented on frontend/src/scenes/max/maxLogic.tsx:
```diff
@@ -70,29 +66,42 @@ export const maxLogic = kea<maxLogicType>([
             // Actions are lazy-loaded. In order to display their names in the UI, we're loading them here.
             actionsModel({ params: 'include_count=1' }),
             ['actions'],
+            router,
+            ['location as routerLocation'],
```
`router` isn't useful to connect, since it's always mounted, it's actually less clear this way I would say. Best to just always directly use `router.selectors.location` or `router.values.location`.

Twixes commented on frontend/src/scenes/max/maxLogic.tsx:
```diff
@@ -70,29 +66,42 @@ export const maxLogic = kea<maxLogicType>([
         resetThread: true,
         cleanThread: true,
         startNewConversation: true,
+        toggleConversationHistory: (visible?: boolean) => ({ visible }),
```
`maxGlobalLogic` would be the better place for this, well, sort of global logic, I think. That's because `maxLogic`, being keyed by thread ID, is largely `maxThreadLogic` (IMO we should actually explicitly refactor it to that name). Hence wary of introducing very non-thread specific state into this logic.

skoob13 replied:
I removed the key from the maxLogic. I don't see a case when a key would be useful anytime soon, do you?

Twixes replied:
It would be useful in a different layout, I agree it's not really useful now. That's actually the key thing: in effect both `maxGlobalLogic` and `maxLogic` are global now.

I'm thinking of a layout like this:
- `maxGlobalLogic` is the one without any key, and it stores the current `conversationId`. This logic would always be mounted when the Max panel or scene is open.
- `maxLogic` becomes `maxThreadLogic` – always keyed by the current `conversationId` of `maxGlobalLogic`. This logic would only be mounted when a specific thread is open, and it would handle only the state of the in-progress thread: loading from history, messages so far, generation of new ones.

But that's new PR material

Twixes reviewed (commented):
Read through the remaining bits to finish off

skoob13 committed:
Revert "feat: wip second back button option"

This reverts commit d3e4668508c76f88682fb6e8000fc17fef4e18bc.

skoob13 committed:
fix: update navigation

skoob13 committed:
fix: router

skoob13 committed:
fix: max on the page

skoob13 committed:
feat: animation

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/max-conversations-ui

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

skoob13 committed:
fix: one line title

skoob13 committed:
fix: ci

skoob13 committed:
Merge branch 'feat/max-conversations-ui' of github.com:PostHog/posthog into feat/max-conversations-ui

skoob13 committed:
fix: handle optional element in getScrollableContainer function

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Make back button animation snappier + prevent button clipping

Twixes committed:
On the home screen, make the title "Max"

Makes it so that we follow the pattern of other side panels, i.e. when you open it, the panel header says the same thing as the panel's button.

Twixes committed:
Polish up skeleton & overflow experience

Twixes committed:
Tweak some margins

Twixes committed:
Wrap SidePanel PanelContent in ErrorBoundary

There's an error when an in-generation thread is exited sometimes, but the bigger issue was that that crashed the WHOLE app. Now it doesn't.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Twixes reviewed (approved):
Pushed a few small UI tweaks, so take a look at those. This is sweet! Let's get it in

The only bug I've caught: If you try to back out of an in-progress conversation, sometimes Max crashes with this error:
```
TypeError: Cannot read properties of undefined (reading 'type')
    at maxLogic.threadGrouped (http://localhost:8234/static/chunk-OCLPHRM4.js:804:54)
    at recomputationWrapper (http://localhost:8234/static/chunk-F5BCIEHM.js:2636:29)
    at memoized (http://localhost:8234/static/chunk-F5BCIEHM.js:2550:24)
    at dependenciesChecker (http://localhost:8234/static/chunk-F5BCIEHM.js:2644:44)
    at Object.memoized [as threadGrouped] (http://localhost:8234/static/chunk-F5BCIEHM.js:2550:24)
    at http://localhost:8234/static/chunk-F5BCIEHM.js:4356:40
    at dependenciesChecker (http://localhost:8234/static/chunk-F5BCIEHM.js:2642:41)
    at Object.memoized [as threadVisible] (http://localhost:8234/static/chunk-F5BCIEHM.js:2550:24)
    at http://localhost:8234/static/chunk-F5BCIEHM.js:4356:40
    at http://localhost:8234/static/chunk-F5BCIEHM.js:3528:16
```
This doesn't seem like a showstopper necessarily, so for now I just ensured content of each side panel is wrapped in an ErrorBoundary, so that we avoid the whole app crashing. This makes the issue very contained. We should still resolve it though.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
Merge branch 'master' into feat/max-conversations-ui

skoob13 committed:
fix: tests

skoob13 committed:
Merge branch 'master' into feat/max-conversations-ui

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

skoob13 committed:
Merge branch 'master' into feat/max-conversations-ui

skoob13 committed:
fix: snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (9)
